### PR TITLE
Preload torchvision during warmup for non-end2end NMS path

### LINKED ISSUE #25022: predict() NMS ~5× slower on 8.4.81+ — fast torchvision.ops.nms path is gated on "torchvision" in sys.modules, which lazy imports no longer satisfy
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Between ultralytics **8.4.80 and 8.4.84**, `predict()` NMS became **~5× slower** for high-candidate-count workloads, with **no change to the NMS code itself**. The NMS dispatch selects the fast CUDA `torchvision.ops.nms` **only if `"torchvision" in sys.modules`**, otherwise a pure-torch fallback (`TorchNMS`, added ~8.3.190). In 8.4.81+ the torchvision imports were made lazy/scoped (the same laziness that fixed SAM/optional-import startup), so during a plain `YOLO(...).predict()` **torchvision is not in `sys.modules`** → the gate silently picks the slow pure-torch path.

**Measurements** — same TensorRT `.engine`, same container/hardware, only the ultralytics wheel swapped (Jetson Orin Nano, JetPack 7.2; the mechanism is hardware-independent):

| ultralytics | NMS stage | throughput |
|---|---|---|
| 8.4.80 | ~4 ms | 66.6 FPS |
| 8.4.84 | ~22 ms | 30.7 FPS |
| 8.4.84 **+ `import torchvision`** | ~4 ms | 65.5 FPS |

Workload: `yolov8n` @ `imgsz=640` fed a low confidence floor (`conf≈0.1` → hundreds of candidate boxes) — exactly where the pure-torch NMS is worst.

**Proof it's the `sys.modules` gate:** forcing `import torchvision` (caller top / `sitecustomize`) on 8.4.84 drops the NMS stage **22→4 ms** and restores **65.5 FPS**, matching 8.4.80. No other change.

### Environment

ultralytics 8.4.84 (and any 8.4.81+); torch 2.x + torchvision (CUDA); Python 3.12; reproduces on any platform (the fallback is pure-torch, not Jetson-specific).

### Minimal Reproducible Example

```python
import time, torch
from ultralytics import YOLO        # do NOT import torchvision here
m = YOLO("yolov8n.pt")
im = "https://ultralytics.com/images/bus.jpg"
for _ in range(30):
    t = time.time()
    m.predict(im, conf=0.05, imgsz=640, verbose=False)  # low conf -> many candidates
    torch.cuda.synchronize()
    print(f"{(time.time()-t)*1000:.1f} ms")
# Add `import torchvision` at the top and re-run: NMS/predict is ~5x faster.
```

### Additional

**Suggested fix**

Don't gate the fast path on `"torchvision" in sys.modules`. Attempt `import torchvision` (try/except) at the point NMS is dispatched — or import it in the NMS module — so the fast CUDA path is used whenever torchvision is installed, regardless of whether some other code path imported it first. The `sys.modules` check is fragile to precisely the lazy-import refactors being done.

**Related:** #19747 (torchvision must be imported before a TorchScript-NMS model loads — same gate, different symptom) and the SAM lazy-import change (which surfaced this). #5059 / #18924 / #22069 are a *separate* `torchvision::nms` CUDA-backend failure, not this slow-path selection.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Fixes #25022 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves inference warmup behavior by preloading `torchvision` for standard NMS paths, helping speed up later non-max suppression calls 🚀

### 📊 Key Changes
- Added a conditional `torchvision` import during `AutoBackend.warmup()` when the model is **not end-to-end**.
- Updated NMS comments to clarify the strategy:
  - Use `torchvision` NMS when it has already been imported.
  - Fall back to `TorchNMS` when `torchvision` is not loaded to avoid slow import overhead.
- Keeps end-to-end models unaffected, since they do not rely on the same post-processing NMS path.

### 🎯 Purpose & Impact
- Reduces first-inference latency after warmup by ensuring `torchvision` NMS is available when needed ⚡
- Preserves low-latency prediction behavior by avoiding unnecessary `torchvision` imports outside warmup.
- Improves consistency across validation, streaming, and warmed-up inference workflows.
- Provides a small but useful performance optimization with no expected user-facing API changes ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution has been received, and an Ultralytics engineer will assist with review soon. To ensure a seamless integration of your warmup/NMS path improvement, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The PR is clean. The added conditional preload is narrowly scoped to non-end2end warmup and does not introduce any evident correctness, performance, or maintainability issues from the visible diff.

**Comment by UltralyticsAssistant** [comment]
🚀 Merged — thank you @Y-T-G for the thoughtful optimization, with contributions from @glenn-jocher!

“Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR is a great example of that principle in action: preloading `torchvision` during `AutoBackend.warmup()` for standard NMS paths is a focused change that helps reduce first-inference latency after warmup while preserving efficient behavior for end-to-end models. The clarified NMS strategy also makes the code easier to reason about across validation, streaming, and warmed-up inference workflows.

Appreciate the attention to performance details and user experience — small improvements like this help make Ultralytics inference faster and more consistent for everyone.
